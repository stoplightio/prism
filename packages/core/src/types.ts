/**
 * The below few interfaces (until end comment) should be defined in the @stoplight/types repo
 */

export interface IDisposable {
  dispose: () => void;
}

export type Path = Array<string | number>;

export enum ValidationSeverity {
  ERROR = 'error',
  WARN = 'warn',
  INFO = 'info',
}
export interface IValidation {
  path: Path;
  name: string;
  summary: string;
  severity: ValidationSeverity;
  message: string;
}

// END

export interface IPrism<Resource, Input, Output, Config, LoadOpts> {
  process: (input: Input, config?: Config) => Promise<IPrismOutput<Input, Output>>;
  load: (opts?: LoadOpts) => Promise<void>;
  readonly resources: Resource[];
}

export interface IPrismConfig {
  mock?: boolean | object;
  security?: boolean | object;
  validate?: boolean | object;
}

export type PrismConfigFactory<C, I> = (input: I, defaultConfig?: PrismConfig<C, I>) => C;
export type PrismConfig<C, I> = C | PrismConfigFactory<C, I>;

export interface ILoader<Options, Resource> {
  load: (opts?: Options, defaultLoader?: ILoader<Options, Resource>) => Promise<Resource[]>;
}

export interface IFilesystemLoaderOpts {
  path?: string;
}

export interface IHttpLoaderOpts {
  url?: string;
}

export interface IRouter<Resource, Input, Config> {
  route: (
    opts: { resources: Resource[]; input: Input; config?: Config },
    defaultRouter?: IRouter<Resource, Input, Config>
  ) => Resource;
}

export interface IForwarder<Resource, Input, Config, Output> {
  forward: (
    opts: { resource?: Resource; input: IPrismInput<Input>; config?: Config },
    defaultForwarder?: IForwarder<Resource, Input, Config, Output>
  ) => Promise<Output>;
}

export interface IMocker<Resource, Input, Config, Output> {
  mock: (
    opts: Partial<IMockerOpts<Resource, Input, Config>>,
    defaultMocker?: IMocker<Resource, Input, Config, Output>
  ) => Promise<Output>;
}

export interface IMockerOpts<Resource, Input, Config> {
  resource: Resource;
  input: IPrismInput<Input>;
  config: Config;
}

export interface IValidator<Resource, Input, Config, Output> {
  validateInput?: (
    opts: { resource: Resource; input: Input; config?: Config },
    defaultValidator?: IValidator<Resource, Input, Config, Output>
  ) => Promise<IValidation[]>;
  validateOutput?: (
    opts: { resource: Resource; output?: Output; config?: Config },
    defaultValidator?: IValidator<Resource, Input, Config, Output>
  ) => Promise<IValidation[]>;
}

export interface IPrismComponents<Resource, Input, Output, Config, LoadOpts> {
  config: PrismConfig<Config, Input>;
  loader: ILoader<LoadOpts, Resource>;
  router: IRouter<Resource, Input, Config>;
  forwarder: IForwarder<Resource, Input, Config, Output>;
  mocker: IMocker<Resource, Input, Config, Output>;
  validator: IValidator<Resource, Input, Config, Output>;
}

export interface IPrismInput<I> {
  data: I;
  validations: {
    input: IValidation[];
  };
}

export interface IPrismOutput<I, O> {
  input?: I;
  output?: O;
  validations: {
    input: IValidation[];
    output: IValidation[];
  };
}
