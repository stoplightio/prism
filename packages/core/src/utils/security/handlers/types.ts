export type SecurityScheme = { type: string; name: string; in?: string; scheme?: string };
