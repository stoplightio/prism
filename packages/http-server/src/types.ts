import { IHttpConfig, PickRequired, PrismHttpComponents, PrismHttpInstance } from '@stoplight/prism-http';
import { FastifyInstance } from 'fastify';

export interface IPrismHttpServerOpts {
  components: PickRequired<Partial<PrismHttpComponents>, 'logger'>;
  config: IHttpConfig;
  cors: boolean;
  errors: boolean;
}

export interface IPrismHttpServer {
  readonly prism: PrismHttpInstance;
  readonly fastify: FastifyInstance;
  listen: (port: number, address?: string, backlog?: number) => Promise<string>;
}
