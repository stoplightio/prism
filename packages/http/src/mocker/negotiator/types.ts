import { IHttpOperationConfig } from '@stoplight/prism-http';
import { INodeExample, INodeExternalExample, ISchema } from '@stoplight/types';

export interface IHttpNegotiationResult {
  code: string;
  mediaType: string;
  example?: INodeExample | INodeExternalExample;
  schema?: ISchema;
}

export type NegotiationOptions = IHttpOperationConfig;

export type NegotiatePartialOptions = {
  code: string;
  dynamic: boolean;
  exampleKey?: string;
};
