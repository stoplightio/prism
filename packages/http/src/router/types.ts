import { IHttpOperationEx } from '../types';

export type Nullable<T> = T | null;

export interface IMatch {
  resource: IHttpOperationEx;
  pathMatch: MatchType;
  methodMatch: MatchType;
  serverMatch?: Nullable<MatchType>;
}

export enum MatchType {
  CONCRETE = 'concrete',
  TEMPLATED = 'templated',
  NOMATCH = 'no-match',
}
