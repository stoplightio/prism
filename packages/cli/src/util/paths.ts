import { generate } from '@stoplight/prism-http/src/mocker/generator/HttpParamGenerator';
import { serializeWithDeepObjectStyle } from '@stoplight/prism-http/src/mocker/serializer/style/deepObject';
import {
  serializeWithCommaDelimitedStyle,
  serializeWithPipeDelimitedStyle,
  serializeWithSpaceDelimitedStyle,
} from '@stoplight/prism-http/src/mocker/serializer/style/delimited';
import {
  HttpParamStyles,
  IHttpOperation,
  IHttpParam,
  IHttpPathParam,
  IHttpQueryParam,
  INodeExample,
  INodeExternalExample,
} from '@stoplight/types';
import * as Either from 'fp-ts/lib/Either';
import * as Option from 'fp-ts/lib/Option';
import { pipe } from 'fp-ts/lib/pipeable';
import { get } from 'lodash';
// @ts-ignore
const { parse } = require('uri-template');

function generateParamValues(specs: IHttpParam[]) {
  return specs.reduce((values, spec) => {
    const value = Option.toUndefined(generate(spec));
    switch (spec.style) {
      case HttpParamStyles.DeepObject:
        values[spec.name] = serializeWithDeepObjectStyle(spec.name, value);
        break;

      case HttpParamStyles.PipeDelimited:
        values[spec.name] = serializeWithPipeDelimitedStyle(
          spec.name,
          value as Array<string | number | boolean>,
          spec.explode,
        );
        break;

      case HttpParamStyles.SpaceDelimited:
        values[spec.name] = serializeWithSpaceDelimitedStyle(
          spec.name,
          value as Array<string | number | boolean>,
          spec.explode,
        );
        break;

      default:
        values[spec.name] = value;
    }
    return values;
  }, {});
}

export function createExamplePath(operation: IHttpOperation): Either.Either<Error, string> {
  return pipe(
    Either.tryCatch(() => {
      const specs = get(operation, 'request.path', []);
      return { template: createPathUriTemplate(operation.path, specs), values: generateParamValues(specs) };
    }, Either.toError),
    Either.chain(({ template, values }) => {
      const specs = get(operation, 'request.query', []);
      try {
        return Either.right({
          template: createQueryUriTemplate(template, specs),
          values: { ...values, ...generateParamValues(specs) },
        });
      } catch (e) {
        return Either.left(e);
      }
    }),
    Either.map(({ template, values }) => parse(template).expand(values)),
  );
}

function createPathUriTemplate(path: string, specs: IHttpPathParam[]) {
  // defaults for query: style=Simple exploded=false
  return specs.reduce(
    (p, spec) =>
      path.replace(
        `{${spec.name}}`,
        createParamUriTemplate(spec.name, spec.style || HttpParamStyles.Simple, spec.explode || false),
      ),
    path,
  );
}

function createQueryUriTemplate(path: string, specs: IHttpQueryParam[]) {
  // defaults for query: style=Form exploded=false
  const formSpecs = specs.filter(spec => (spec.style || HttpParamStyles.Form) === HttpParamStyles.Form);
  const formExplodedParams = formSpecs
    .filter(spec => spec.explode)
    .map(spec => spec.name)
    .join(',');
  const formImplodedParams = formSpecs
    .filter(spec => !spec.explode)
    .map(spec => spec.name)
    .join(',');
  const restParams = specs
    .filter(spec =>
      [HttpParamStyles.DeepObject, HttpParamStyles.SpaceDelimited, HttpParamStyles.PipeDelimited].includes(spec.style),
    )
    .map(spec => spec.name)
    .map(name => `{+${name}}`)
    .join('&');

  if (formExplodedParams) {
    path += `{?${formExplodedParams}*}`;
  }

  if (formImplodedParams) {
    path += `{${formExplodedParams ? '&' : '?'}${formImplodedParams}}`;
  }

  if (restParams) {
    path += `${formExplodedParams || formImplodedParams ? '&' : '?'}${restParams}`;
  }

  return path;
}

function createParamUriTemplate(name: string, style: HttpParamStyles, explode: boolean) {
  const starOrVoid = explode ? '*' : '';
  switch (style) {
    case HttpParamStyles.Simple:
      return `{${name}${starOrVoid}}`;
    case HttpParamStyles.Label:
      return `{.${name}${starOrVoid}}`;
    case HttpParamStyles.Matrix:
      return `{;${name}${starOrVoid}}`;
    default:
      throw new Error(`Unsupported parameter style: ${style}`);
  }
}
