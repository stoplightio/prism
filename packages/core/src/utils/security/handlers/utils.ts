import { DiagnosticSeverity } from '@stoplight/types';
import { Either, left, right } from 'fp-ts/lib/Either';
import { IPrismDiagnostic } from '../../../types';

const forbiddenErr: IPrismDiagnostic = {
  code: 403,
  message: 'Invalid credentials used',
  severity: DiagnosticSeverity.Error,
};

const invalidCredsErr = left(forbiddenErr);

export function genRespForScheme<R>(
  isSchemeProper: boolean,
  isCredsGiven: boolean,
  resource: R,
  msg: string,
): Either<IPrismDiagnostic, R> {
  const handler = [
    {
      test: () => isSchemeProper && isCredsGiven,
      handle: () => right(resource),
    },
    {
      test: () => isSchemeProper,
      handle: () => invalidCredsErr,
    },
  ].find(possibleHandler => possibleHandler.test());

  return handler ? handler.handle() : left(genUnauthorisedErr(msg));
}

export const genUnauthorisedErr = (msg: string): IPrismDiagnostic => ({
  severity: DiagnosticSeverity.Error,
  message: 'Invalid security scheme used',
  code: 401,
  tags: msg ? [msg] : [],
});

export function isScheme(shouldBeScheme: string, authScheme: string) {
  return (authScheme || '').toLowerCase() === shouldBeScheme;
}

export function when(condition: boolean, errorMessage: string, resource?: unknown): Either<IPrismDiagnostic, unknown> {
  return condition ? right(resource) : left(genUnauthorisedErr(errorMessage));
}
