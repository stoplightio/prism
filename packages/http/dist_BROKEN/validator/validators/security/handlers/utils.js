"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.when = exports.isScheme = exports.genUnauthorisedErr = exports.genRespForScheme = void 0;
const types_1 = require("@stoplight/types");
const E = require("fp-ts/Either");
const function_1 = require("fp-ts/function");
const function_2 = require("fp-ts/function");
const genRespForScheme = (isSchemeProper, isCredsGiven, msg) => function_1.pipe(isSchemeProper, E.fromPredicate(function_2.identity, () => exports.genUnauthorisedErr(msg)), E.chain(() => exports.when(isCredsGiven, undefined)));
exports.genRespForScheme = genRespForScheme;
const genUnauthorisedErr = (msg) => ({
    severity: types_1.DiagnosticSeverity.Error,
    message: 'Invalid security scheme used',
    code: 401,
    tags: msg ? [msg] : [],
});
exports.genUnauthorisedErr = genUnauthorisedErr;
function isScheme(shouldBeScheme, authScheme) {
    return authScheme.toLowerCase() === shouldBeScheme;
}
exports.isScheme = isScheme;
const when = (condition, errorMessage) => function_1.pipe(condition, E.fromPredicate(function_2.identity, () => exports.genUnauthorisedErr(errorMessage)));
exports.when = when;
