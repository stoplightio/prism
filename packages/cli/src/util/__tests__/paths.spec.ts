import { HttpParamStyles } from '@stoplight/types';
import { createExamplePath } from '../paths';
import { assertRight, assertLeft } from '@stoplight/prism-core/src/__tests__/utils';

describe('createExamplePath()', () => {
  describe('path parameters', () => {
    it('generates simple style', () => {
      assertRight(
        createExamplePath({
          id: '123',
          path: '/path/{p}',
          method: 'get',
          request: { path: [{ name: 'p', style: HttpParamStyles.Simple, examples: [{ key: 'foo', value: 'test' }] }] },
          responses: [{ code: '200' }],
        }),
        r => expect(r).toEqual('/path/test')
      );
    });

    it('generates label style', () => {
      assertRight(
        createExamplePath({
          id: '123',
          path: '/path/{p}',
          method: 'get',
          request: { path: [{ name: 'p', style: HttpParamStyles.Label, examples: [{ key: 'foo', value: 'test' }] }] },
          responses: [{ code: '200' }],
        }),
        r => expect(r).toEqual('/path/.test')
      );
    });

    it('generates matrix style', () => {
      assertRight(
        createExamplePath({
          id: '123',
          path: '/path/{p}',
          method: 'get',
          request: { path: [{ name: 'p', style: HttpParamStyles.Matrix, examples: [{ key: 'foo', value: 'test' }] }] },
          responses: [{ code: '200' }],
        }),
        r => expect(r).toEqual('/path/;p=test')
      );
    });
  });

  describe('query parameters', () => {
    it('generates form style', () => {
      assertRight(
        createExamplePath({
          id: '123',
          path: '/path',
          method: 'get',
          request: { query: [{ name: 'p', style: HttpParamStyles.Form, examples: [{ key: 'foo', value: 'test' }] }] },
          responses: [{ code: '200' }],
        }),
        r => expect(r).toEqual('/path?p=test')
      );
    });

    it('generates deepObject style', () => {
      assertRight(
        createExamplePath({
          id: '123',
          path: '/path',
          method: 'get',
          request: {
            query: [
              {
                name: 'p',
                style: HttpParamStyles.DeepObject,
                examples: [{ key: 'foo', value: { a: { aa: 1, ab: 2 } } }],
              },
            ],
          },
          responses: [{ code: '200' }],
        }),
        r => expect(r).toEqual('/path?p%5Ba%5D%5Baa%5D=1&p%5Ba%5D%5Bab%5D=2')
      );
    });

    it('generates pipeDelimited style', () => {
      assertRight(
        createExamplePath({
          id: '123',
          path: '/path',
          method: 'get',
          request: {
            query: [
              {
                name: 'p',
                style: HttpParamStyles.PipeDelimited,
                examples: [{ key: 'foo', value: [1, 2, 3] }],
              },
            ],
          },
          responses: [{ code: '200' }],
        }),
        r => expect(r).toEqual('/path?p=1%7C2%7C3')
      );
    });

    it('generates spaceDelimited style', () => {
      assertRight(
        createExamplePath({
          id: '123',
          path: '/path',
          method: 'get',
          request: {
            query: [
              {
                name: 'p',
                style: HttpParamStyles.SpaceDelimited,
                examples: [{ key: 'foo', value: [1, 2, 3] }],
              },
            ],
          },
          responses: [{ code: '200' }],
        }),
        r => expect(r).toEqual('/path?p=1%202%203')
      );
    });

    it('fails when invalid example provided for pipeDelimited style', () => {
      assertLeft(
        createExamplePath({
          id: '123',
          path: '/path',
          method: 'get',
          request: {
            query: [{ name: 'q', style: HttpParamStyles.PipeDelimited, examples: [{ key: 'foo', value: 'test' }] }],
          },
          responses: [{ code: '200' }],
        }),
        e => expect(e.message).toEqual('Pipe delimited style is only applicable to array parameter')
      );
    });

    it('fails when invalid example provided for spaceDelimited style', () => {
      assertLeft(
        createExamplePath({
          id: '123',
          path: '/path',
          method: 'get',
          request: {
            query: [{ name: 'q', style: HttpParamStyles.SpaceDelimited, examples: [{ key: 'foo', value: 'test' }] }],
          },
          responses: [{ code: '200' }],
        }),
        e => expect(e.message).toEqual('Space delimited style is only applicable to array parameter')
      );
    });
  });

  describe('mixed parameters', () => {
    it('generates correct path', () => {
      assertRight(
        createExamplePath({
          id: '123',
          path: '/path/{p1}/{p2}/{p3}',
          method: 'get',
          request: {
            path: [
              { name: 'p1', style: HttpParamStyles.Simple, examples: [{ key: 'foo', value: 'test1' }] },
              { name: 'p2', style: HttpParamStyles.Label, examples: [{ key: 'foo', value: ['test1', 'test2'] }] },
              { name: 'p3', style: HttpParamStyles.Matrix, examples: [{ key: 'foo', value: ['test1', 'test2'] }] },
            ],
            query: [
              { name: 'q1', style: HttpParamStyles.Form, examples: [{ key: 'foo', value: 'test1' }] },
              {
                name: 'q2',
                style: HttpParamStyles.SpaceDelimited,
                examples: [{ key: 'foo', value: ['test1', 'test2'] }],
              },
              {
                name: 'q3',
                style: HttpParamStyles.PipeDelimited,
                examples: [{ key: 'foo', value: ['test1', 'test2'] }],
              },
              {
                name: 'q4',
                style: HttpParamStyles.PipeDelimited,
                explode: true,
                examples: [{ key: 'foo', value: ['test1', 'test2'] }],
              },
              {
                name: 'q5',
                style: HttpParamStyles.DeepObject,
                examples: [{ key: 'foo', value: { a: ['test1', 'test2'], b: { ba: 1, bb: 2 } } }],
              },
            ],
          },
          responses: [{ code: '200' }],
        }),
        r =>
          expect(r).toEqual(
            '/path/test1/.test1,test2/;p3=test1,test2?q1=test1&q2=test1%20test2&q3=test1%7Ctest2&q4=test1&q4=test2&q5%5Ba%5D%5B%5D=test1&q5%5Ba%5D%5B%5D=test2&q5%5Bb%5D%5Bba%5D=1&q5%5Bb%5D%5Bbb%5D=2'
          )
      );
    });
  });
});
