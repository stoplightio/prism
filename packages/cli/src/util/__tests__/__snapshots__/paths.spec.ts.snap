// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createExamplePath() path parameters fails when not supported style supplied 1`] = `[Error: Unsupported parameter style: pipeDelimited]`;
